crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2398,32 +2377,58 @@ impl AuthorityPerEpochStore {
                     );
                     let mut lock = self.get_reconfig_state_write_lock_guard();
                     lock.close_all_certs();
-                    // We store reconfig_state and end_of_publish in same batch to avoid dealing with inconsistency here on restart
                     self.store_reconfig_state_batch(&lock, write_batch)?;
-                    write_batch.insert_batch(
-                        &self.tables.final_epoch_checkpoint,
-                        [(
-                            &FINAL_EPOCH_CHECKPOINT_INDEX,
-                            &consensus_index.last_committed_round,
-                        )],
-                    )?;
                     // Holding this lock until end of this function where we write batch to DB
-                    ret = Some((lock, consensus_index.last_committed_round));
+                    ret = Some(lock);
                 };
-                // Important: we actually rely here on fact that ConsensusHandler panics if it's
+                // Important: we actually rely here on fact that ConsensusHandler panics if its
                 // operation returns error. If some day we won't panic in ConsensusHandler on error
                 // we need to figure out here how to revert in-memory state of .end_of_publish
                 // and .reconfig_state when write fails.
                 self.record_consensus_message_processed(write_batch, transaction.key())?;
             } else {
                 panic!(
-                    "process_end_of_publish_transaction called with non-end-of-publish transaction"
+                    "process_end_of_publish_transactions called with non-end-of-publish transaction"
                 );
             }
         }
         Ok(ret)
     }
 
+    fn process_end_of_all_transactions<'a>(
+        &'a self,
+        write_batch: &mut DBBatch,
+        lock: Option<parking_lot::RwLockWriteGuard<'a, ReconfigState>>,
+        commit_has_deferred_txns: bool,
+    ) -> SuiResult<(
+        Option<parking_lot::RwLockWriteGuard<'a, ReconfigState>>,
+        bool,
+    )> {
+        let is_reject_all_certs = if let Some(lock) = &lock {
+            lock.is_reject_all_certs()
+        } else {
+            // It is ok to just release lock here as this function is the only place that
+            // transitions into RejectAllTx state, and this function itself is always
+            // executed from consensus task.
+            self.get_reconfig_state_read_lock_guard()
+                .is_reject_all_certs()
+        };
+
+        if !is_reject_all_certs || !self.deferred_transactions_empty() || commit_has_deferred_txns {
+            // Don't end epoch until all deferred transactions are processed.
+            return Ok((lock, false));
+        }
+
+        let mut lock = if let Some(lock) = lock {
```

Comment:
nit: this might be more concise as an `unwrap_or_else`

Reply:
done